### DISCUSSION TITLE
Yolov8 not training properly when using gpu

### DISCUSSION THREAD

**Discussion by Ziad-Assem** [ultralytics/ultralytics#4455] [Q&A]
hello, im trying to train a yolov8 model but when it's being trained, it prints nan at every parameter and the model ends up not detecting anything. please help:

output i get:

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       1/20      2.28G        nan        nan        nan         38        640: 100%|██████████| 5/5 [00:18<00:00,  3.65s/it]
D:\Program Files\Python39\torch\optim\lr_scheduler.py:139: UserWarning: Detected call of `lr_scheduler.step()` before `optimizer.step()`. In PyTorch 1.1.0 and later, you should call them in the opposite order: `optimizer.step()` before `lr_scheduler.step()`.  Failure to do this will result in PyTorch skipping the first value of the learning rate schedule. See more details at https://pytorch.org/docs/stable/optim.html#how-to-adjust-learning-rate
  warnings.warn("Detected call of `lr_scheduler.step()` before `optimizer.step()`. "
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  1.55it/s]
                   all         20         20          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       2/20      2.27G        nan        nan        nan         48        640: 100%|██████████| 5/5 [00:03<00:00,  1.44it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  3.45it/s]
                   all         20         20          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       3/20      2.27G        nan        nan        nan         38        640: 100%|██████████| 5/5 [00:03<00:00,  1.44it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  3.53it/s]
                   all         20         20          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       4/20      2.28G        nan        nan        nan         27        640: 100%|██████████| 5/5 [00:03<00:00,  1.47it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  3.62it/s]
                   all         20         20          0          0          0          0

      Epoch    GPU_mem   box_loss   cls_loss   dfl_loss  Instances       Size
       5/20      2.28G        nan        nan        nan         36        640: 100%|██████████| 5/5 [00:03<00:00,  1.47it/s]
                 Class     Images  Instances      Box(P          R      mAP50  mAP50-95): 100%|██████████| 1/1 [00:00<00:00,  3.62it/s]
                   all         20         20          0          0          0          0

code i use:

from ultralytics import YOLO
import multiprocessing as mp

# Load a model
model = YOLO('yolov8n.yaml')  # build a new model from YAML

# Train the model
if __name__ == '__main__':
    # Tell Python to use the fork method to start child processes.
    mp.freeze_support()

    results = model.train(data=rf"D:\Model Training\testing playground\datasets\C Image Dataset\data.yaml", epochs=20, project="vscode_test", device=0)

**Comment by glenn-jocher** [comment]
@Ziad-Assem hello! It looks like you're encountering NaN (not a number) values during training, which typically indicates that the model's loss has diverged. This can happen for several reasons, such as learning rate being too high, improper data normalization, or issues with the dataset labels.

Here are a few steps you can take to troubleshoot the issue:

1. **Check Dataset**: Ensure that your dataset is correctly labeled and that the annotations are in the proper format. Also, verify that there are no corrupt images or annotations.

2. **Learning Rate**: Try reducing the learning rate. A high learning rate can cause the model to diverge during the initial stages of training.

3. **Batch Size**: If possible, increase the batch size to stabilize training, but make sure it fits within your GPU memory limits.

4. **Data Normalization**: Confirm that your data is being normalized correctly. The input images should typically be normalized to have values between 0 and 1 or -1 and 1.

5. **Pretrained Weights**: Start training with pretrained weights if you aren't already. This can provide a more stable starting point for the training process.

6. **Gradient Clipping**: Implement gradient clipping to prevent gradients from exploding, which can lead to NaN losses.

7. **Optimizer and Scheduler**: Make sure that the optimizer and learning rate scheduler are configured correctly. As the warning suggests, the order of `optimizer.step()` and `lr_scheduler.step()` is important.

8. **Environment**: Ensure that your training environment is set up correctly with all necessary dependencies and that you're using compatible versions of PyTorch and other libraries.

If you've gone through these steps and the issue persists, you might want to look into the training logs for more detailed insights or consider reaching out to the community for support. Remember, the Ultralytics documentation can be a valuable resource for understanding the expected configurations and usage patterns. Keep experimenting, and good luck with your training! 🚀
